### PR TITLE
Complete transition of a second (or greater) failed primary

### PR DESCRIPTION
**Pull request by gkokolatos**
Assume that there exists a healthy three datanode cluster having datanodes named
node1, node2, node3. Assume that node1 is the primary and it fails. The cluster
will correctly demote node1 and promote node2 to primary. When node2 fails, the
cluster would fail to reach a stable state due to node2 getting stuck in
draining state.

The culrpit is GetPrimaryOrDemotedNodeInGroup which contains logic to try to
find a demoted primary in case that a healty primary is not existing. To do
that, it traverses an ordered by name list of nodes in the group. When it finds
the first node that is either a demoted or draining or taken out for
maintainance primary, it chooses it and stops looking further. In the example
above, note that the ordered by name list starts with node1, which would satisfy
the criterion and node2 will never be selected.

To fix that, continue looping through the nodes. Assuming that there can be only
one primary active at any time, hence only one primary that is currently going
through demotion process, then choose that one.

Add a specific test case for it.

Fixes issue #704

### CODE REVIEW

File: tests/test_multinode_alternate_primary_failures.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+def test_005_001_fail_primary_again():
+    # verify that node3 is primary and stop it
+    assert node3.get_state().assigned == "primary"
+    node3.fail()
+
+    assert node3.wait_until_assigned_state(target_state="demoted")
+    assert node1.wait_until_assigned_state(target_state="primary")
+
+def test_005_002_fail_primary_again():
+    # verify that node3 is primary and stop it
```
Comment (DimCitus) [inline tests/test_multinode_alternate_primary_failures.py:125]: copy/paste'o, I think that should read “verify that node1 is primary and stop it”

File: src/monitor/node_metadata.c [hunk 2]
```diff
@@ -438,15 +438,20 @@ GetPrimaryOrDemotedNodeInGroup(char *formationId, int32 groupId)
...
+	 * currently being demoted.
+	 */
 	foreach(nodeCell, groupNodeList)
 	{
 		AutoFailoverNode *currentNode = (AutoFailoverNode *) lfirst(nodeCell);
 
-		if (StateBelongsToPrimary(currentNode->reportedState))
+		if (StateBelongsToPrimary(currentNode->reportedState) &&
+			(IsBeingDemotedPrimary(primaryNode) == false ||
+			IsDemotedPrimary(currentNode) == false))
```
Comment (DimCitus) [inline src/monitor/node_metadata.c:452]: I think we would generally use a slightly different style:

``` c
		if (StateBelongsToPrimary(currentNode->reportedState) &&
  			(!IsBeingDemotedPrimary(primaryNode) ||
  			!IsDemotedPrimary(currentNode)))
```

File: src/monitor/node_metadata.c [hunk 3]
```diff
@@ -438,15 +438,20 @@ GetPrimaryOrDemotedNodeInGroup(char *formationId, int32 groupId)
...
 		{
 			primaryNode = currentNode;
-			break;
```
Comment (DimCitus) [inline src/monitor/node_metadata.c:449]: Do we need to remove the break here? My understanding is that a short cut in this loop wouldn't have an impact, given the new tests, we find only one node and may return either the first or the last one then, it should be the same node, right?

**Review by DimCitus — APPROVED** [review]
Thanks for this bug fix! I like how limited your code changes are.

Please review https://travis-ci.com/github/citusdata/pg_auto_failover/jobs/505381364 and run a `make indent` pass to make Travis happy, and then I'll be able to merge your PR. I have left also some comments, those are not stoppers but you might still want to consider them.

File: tests/test_multinode_alternate_primary_failures.py (hunk 1 above)
Comment (gkokolatos) [reply tests/test_multinode_alternate_primary_failures.py:125]: You are correct, will fix.

File: src/monitor/node_metadata.c (hunk 2 above)
Comment (gkokolatos) [reply src/monitor/node_metadata.c:452]: Heh, I was debating as to which was preferable :) Will fix.

File: src/monitor/node_metadata.c (hunk 3 above)
Comment (gkokolatos) [reply src/monitor/node_metadata.c:449]: I have not tested it with the break to be honest. My understanding is the same as yours. I chose to keep the full loop under a 'belts and suspenders' argument, helped by the fact that the loops are generally very short anyways.

**Comment by gkokolatos** [comment]
Will push a couple of commits shortly with the ident run (always forget that one) and comment derived touches.

**Comment by gkokolatos** [comment]
> Will push a couple of commits shortly with the ident run (always forget that one) and comment derived touches.

Done

**Comment by DimCitus** [comment]
Also, I think we want your new test to be part of the automated test suite on Travis, and for that we need to manually edit the Makefile to add the file in the TESTS_MULTI list. (Beware that I clicked on “update branch” on Travis so your local version will have diverged and you might want to rebase / force push when you do more changes to it).

**Comment by gkokolatos** [comment]
‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
On Friday, 14 May 2021 13:10, Dimitri Fontaine ***@***.***> wrote:

> Also, I think we want your new test to be part of the automated test suite on Travis, and for that we need to manually edit the Makefile to add the file in the TESTS_MULTI list. (Beware that I clicked on “update branch” on Travis so your local version will have diverged and you might want to rebase / force push when you do more changes to it).

Sure. Will do!

**Comment by gkokolatos** [comment]
> ‐‐‐‐‐‐‐ Original Message ‐‐‐‐‐‐‐
> On Friday, 14 May 2021 13:10, Dimitri Fontaine ***@***.***> wrote: Also, I think we want your new test to be part of the automated test suite on Travis, and for that we need to manually edit the Makefile to add the file in the TESTS_MULTI list. (Beware that I clicked on “update branch” on Travis so your local version will have diverged and you might want to rebase / force push when you do more changes to it).
> Sure. Will do!

Done and travis now seems happy.

> […](#)

**Pull request merged**